File: 351001/Boukhvalova/module-publisher/src/main/java/by/boukhvalova/distcomp/controllers/v2/NoteV2Controller.java
package by.boukhvalova.distcomp.controllers.v2;

import by.boukhvalova.distcomp.dto.NoteRequestTo;
import by.boukhvalova.distcomp.dto.NoteResponseTo;
import by.boukhvalova.distcomp.entities.UserRole;
import by.boukhvalova.distcomp.security.user.AuthenticatedUser;
import by.boukhvalova.distcomp.services.NoteService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;

@RestController
@RequestMapping({"/api/v2.0/notes", "/api/v2.0/messages", "/api/v2.0/notices", "/api/v2.0/reactions"})
@RequiredArgsConstructor
public class NoteV2Controller {
    private final NoteService noteService;

    @GetMapping
    public Collection<NoteResponseTo> getAll() {
        return noteService.getAll();
    }

    @GetMapping("/{id}")
    public NoteResponseTo getById(@PathVariable Long id) {
        return noteService.getById(id);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public NoteResponseTo create(@RequestBody @Valid NoteRequestTo request, Authentication authentication) {
        AuthenticatedUser user = (AuthenticatedUser) authentication.getPrincipal();
        if (user.getRole() == UserRole.CUSTOMER) {
            request.setUserId(user.getId());
        }
        return noteService.create(request);
    }

    @PutMapping("/{id}")
    @ResponseStatus(HttpStatus.OK)
    @PreAuthorize("hasRole('ADMIN') or @ownershipService.isNoteOwner(#id, authentication)")
    public NoteResponseTo update(
            @PathVariable Long id,
            @RequestBody @Valid NoteRequestTo request,
            Authentication authentication
    ) {
        request.setId(id);
        AuthenticatedUser user = (AuthenticatedUser) authentication.getPrincipal();
        if (user.getRole() == UserRole.CUSTOMER) {
            request.setUserId(user.getId());
        }
        return noteService.update(request);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @PreAuthorize("hasRole('ADMIN') or @ownershipService.isNoteOwner(#id, authentication)")
    public void delete(@PathVariable Long id) {
        noteService.delete(id);
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/controller/NoteV2Controller.kt
package com.example.distcomp.controller

import com.example.distcomp.dto.request.NoteRequestTo
import com.example.distcomp.dto.response.NoteResponseTo
import com.example.distcomp.security.AuthorizationService
import com.example.distcomp.service.NoteService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2.0/notes")
class NoteV2Controller(
    private val service: NoteService,
    private val authorizationService: AuthorizationService
) {
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(@Valid @RequestBody request: NoteRequestTo): NoteResponseTo {
        authorizationService.ensureCanCreateNote(request.tweetId)
        return service.create(request)
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: Long): NoteResponseTo = service.getById(id)

    @GetMapping
    fun getAll(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(defaultValue = "id,desc") sort: Array<String>
    ): List<NoteResponseTo> = service.getAll(page, size, sort)

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @Valid @RequestBody request: NoteRequestTo): NoteResponseTo {
        authorizationService.ensureCanManageNote(id, request.tweetId)
        return service.put(id, request)
    }

    @PatchMapping("/{id}")
    fun patch(@PathVariable id: Long, @Valid @RequestBody request: NoteRequestTo): NoteResponseTo {
        authorizationService.ensureCanManageNote(id, request.tweetId)
        return service.patch(id, request)
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: Long) {
        authorizationService.ensureCanManageNote(id)
        service.delete(id)
    }
}
